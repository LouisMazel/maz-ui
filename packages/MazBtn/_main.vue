<template>
  <button
    :id="uniqueId"
    class="maz-btn btn"
    :class="[
      classes,
      { hidden: loader }
    ]"
    v-bind="$attrs"
    :type="type"
    :disabled="isDisabled"
    @click="handleClick"
  >
    <slot />
    <div
      v-if="loader"
      class="maz-btn__spinner flex align-center justify-center"
    >
      <MazSpinner
        :size="25"
        dark
      />
    </div>
  </button>
</template>

<script>
  import MazSpinner from '../MazSpinner'
  import uniqueId from './../mixins/uniqueId'

  export default {
    name: 'MazBtn',
    components: {
      MazSpinner
    },
    mixins: [uniqueId],
    props: {
      id: { type: String, default: 'MazBtn' },
      color: { type: String, default: 'primary' },
      size: { type: String, default: null },
      type: { type: String, default: 'button' },
      loader: { type: Boolean, default: false },
      disabled: { type: Boolean, default: false },
      outline: { type: Boolean, default: false },
      rounded: { type: Boolean, default: false },
      fab: { type: Boolean, default: false }
    },
    computed: {
      isDisabled () {
        const { disabled, loader } = this
        return loader || disabled
      },
      classes () {
        const { color, size, outline, rounded, isDisabled, fab } = this
        return [
          ...(color ? [`btn--${color}`] : [null]),
          ...(size ? [`btn--${size}`] : [null]),
          ...(outline ? [`btn--${color}--outline`] : [null]),
          ...(rounded ? [`btn--rounded`] : [null]),
          ...(fab ? [`btn--fab`] : [null]),
          ...(isDisabled ? [`btn--disabled`] : [null])
        ]
      }
    },
    methods: {
      handleClick (evt) {
        this.$emit('click', evt)
      }
    }
  }
</script>

<style lang="scss" scoped>
  .maz-btn {
    position: relative;

    &.hidden {
      color: transparent;
    }

    &__spinner {
      position: absolute;
      margin-left: auto;
      margin-right: auto;
      left: 0;
      top: 0;
      right: 0;
      bottom: 0;
    }
  }
</style>
