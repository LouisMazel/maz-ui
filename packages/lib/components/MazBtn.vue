<template>
  <Component
    :is="component"
    :disabled="isDisabled"
    class="m-btn"
    :class="[
      `--${size}`,
      btnColorClass,
      cursorClass,
      variantClass,
      {
        '--block': block,
        '--no-underline': noUnderline,
        '--no-leading': noLeading,
        '--fab': fab,
        '--loading': loading,
        '--disabled': isDisabled,
        '--icon': hasIcon,
        '--rounded': rounded,
        '--no-rounded': noRounded,
        '--no-padding': noPadding,
        '--no-elevation': noElevation,
      },
    ]"
    :type="btnType"
  >
    <div
      v-if="hasLeftIcon"
      class="m-btn__icon-left maz-flex maz-flex-center"
      :class="{ 'maz-invisible': hasLoader }"
    >
      <!--
        @slot left-icon - The icon to display on the left of the button
      -->
      <slot name="left-icon">
        <MazIcon v-if="typeof leftIcon === 'string'" :name="leftIcon" />
        <Component :is="leftIcon" v-else-if="leftIcon" />
      </slot>
    </div>

    <div v-if="hasFabIcon" class="m-btn__icon" :class="{ 'maz-invisible': hasLoader }">
      <!--
        @slot icon - The icon to display on the fab button
      -->
      <slot name="icon">
        <MazIcon v-if="typeof icon === 'string'" :name="icon" />
        <Component :is="icon" v-else-if="icon" />
      </slot>
    </div>

    <span v-if="$slots.default" :class="[{ 'maz-invisible': hasLoader }, contentClass]">
      <!--
        @slot default - The content of the button
      -->
      <slot></slot>
    </span>

    <div v-if="hasRightIcon" class="m-btn__icon-right" :class="{ 'maz-invisible': hasLoader }">
      <!--
        @slot right-icon - The icon to display on the right of the button
      -->
      <slot name="right-icon">
        <MazIcon v-if="typeof rightIcon === 'string'" :name="rightIcon" />
        <Component :is="rightIcon" v-else-if="rightIcon" />
      </slot>
    </div>

    <MazSpinner v-if="hasLoader" class="m-btn-loader" size="2em" :color="color" />
  </Component>
</template>

<script lang="ts" setup>
  import {
    computed,
    useAttrs,
    useSlots,
    defineAsyncComponent,
    onBeforeMount,
    type FunctionalComponent,
    type SVGAttributes,
    type ComponentPublicInstance,
    type Component,
  } from 'vue'

  import type { Color, Size } from './types'
  export type { Color, Size }

  const MazSpinner = defineAsyncComponent(() => import('./MazSpinner.vue'))
  const MazIcon = defineAsyncComponent(() => import('./MazIcon.vue'))

  const { href, to } = useAttrs()
  const slots = useSlots()

  type Icon = FunctionalComponent<SVGAttributes> | ComponentPublicInstance | Component

  const props = withDefaults(
    defineProps<{
      /** The variant of the button - Change UI of component (link or button style)
       * @values `'button' | 'link'`
       * */
      variant?: 'button' | 'link'
      /**
       * The size of the button
       * @values `'xl' | 'lg' | 'md' | 'sm' | 'xs' | 'mini'`
       */
      size?: Size
      /**
       * The color of the button
       * @values `'primary' | 'secondary' | 'info' | 'success' | 'warning' | 'danger' | 'white' | 'black' | 'transparent' | 'theme'`
       */
      color?: Color
      /**
       * The type of the button
       * @values `'submit' | 'reset' | 'button'`
       */
      type?: 'submit' | 'reset' | 'button'
      /**
       * If true, the button will have a full border radius
       * @default false
       */
      rounded?: boolean
      /**
       * If true, the button will have no border radius
       * @default false
       */
      noRounded?: boolean
      /**
       * If true, the button have the "border" style
       * @default false
       */
      outline?: boolean
      /**
       * If true, the button will have a pastel color
       * @default false
       */
      pastel?: boolean
      /**
       * If true, the button will have a full width
       * @default false
       */
      block?: boolean
      /** If true, the button will have no underline on hover (useful with `variant="link"`)
       * @default false
       */
      noUnderline?: boolean
      /**
       * If true, the button will have no leading (useful with `variant="link"`)
       * @default false
       */
      noLeading?: boolean
      /**
       * Enable the button loader
       * @default false
       */
      loading?: boolean
      /**
       * Disable the button
       * @default false
       */
      disabled?: boolean
      /**
       * If true, the button will have a fab style
       * @default false
       */
      fab?: boolean
      /**
       * The name of the icon to display or component, only with fab
       * `@type` `{string | FunctionalComponent<SVGAttributes> | ComponentPublicInstance | Component}`
       */
      icon?: string | Icon
      /**
       * The name of the icon or component to display on the left of the button
       * `@type` `{string | FunctionalComponent<SVGAttributes> | ComponentPublicInstance | Component}`
       */
      leftIcon?: string | Icon
      /**
       * The name of the icon or component to display on the right of the button
       * `@type` `{string | FunctionalComponent<SVGAttributes> | ComponentPublicInstance | Component}`
       */
      rightIcon?: string | Icon
      /**
       * If true, the button will have no padding
       * @default false
       */
      noPadding?: boolean
      /**
       * If true, the button will have no box-shadow
       * @default false
       */
      noElevation?: boolean
      /** The class applied to the content wrapper (<span />) of the button */
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      contentClass?: any
    }>(),
    {
      variant: 'button',
      size: 'md',
      color: 'primary',
      type: 'button',
      icon: undefined,
      leftIcon: undefined,
      rightIcon: undefined,
      contentClass: undefined,
    },
  )

  onBeforeMount(() => {
    if (props.icon && !props.fab) {
      console.error('[maz-ui](MazBtn) the prop "icon" must be used only with "fab" props')
    }
  })

  const component = computed(() => {
    if (href) return 'a'
    else if (to) return 'router-link'
    return 'button'
  })

  const btnColorClass = computed(() =>
    props.pastel
      ? `--${props.color}-pastel`
      : props.outline
        ? `--${props.color}-outline`
        : `--${props.color}`,
  )
  const isDisabled = computed(
    () => (props.loading || props.disabled) && component.value === 'button',
  )
  const cursorClass = computed(() => (isDisabled.value ? '--cursor-default' : '--cursor-pointer'))
  const variantClass = computed(() => `--is-${props.variant}`)
  const hasLoader = computed(() => props.loading && props.variant === 'button')
  const hasLeftIcon = computed(() => !!slots['left-icon'] || props.leftIcon)
  const hasRightIcon = computed(() => !!slots['right-icon'] || props.rightIcon)
  const hasIcon = computed(() => hasLeftIcon.value || hasRightIcon.value)
  const hasFabIcon = computed(() => props.fab && (props.icon || !!slots['icon']))
  const btnType = computed(() => (component.value === 'button' ? props.type : undefined))
</script>

<style lang="postcss" scoped>
  .m-btn {
    @apply maz-flex maz-flex-none maz-items-center maz-gap-2 maz-border maz-border-solid maz-border-transparent maz-text-center maz-text-base maz-text-normal;

    & span {
      @apply maz-leading-none;
    }

    &-loader {
      @apply maz-absolute;
    }

    &.--cursor-pointer {
      @apply maz-cursor-pointer;
    }

    &.--cursor-default {
      @apply maz-cursor-default;
    }

    &.--is-link {
      @apply maz-inline-flex maz-items-center maz-bg-transparent
        maz-outline-none maz-transition maz-duration-200 maz-ease-in-out;

      &:not(.--no-leading) span {
        @apply maz-leading-9;
      }

      &:not(:disabled):hover,
      &:not(:disabled):focus {
        &:not(.--no-underline) {
          @apply maz-underline;
        }
      }

      &.--secondary {
        @apply maz-text-secondary;
      }

      &.--info {
        @apply maz-text-info;
      }

      &.--warning {
        @apply maz-text-warning-600;
      }

      &.--danger {
        @apply maz-text-danger-600;
      }

      &.--success {
        @apply maz-text-success-600;
      }

      &.--white {
        @apply maz-text-white;
      }

      &.--black {
        @apply maz-text-black;
      }

      &.--theme {
        @apply maz-text-normal;
      }
    }

    &.--is-button {
      @apply maz-relative maz-inline-flex maz-items-center maz-justify-center maz-overflow-hidden
        maz-border-transparent maz-bg-transparent maz-font-medium maz-no-underline
        maz-transition maz-duration-300 maz-ease-in-out;

      &:not(.--no-rounded) {
        @apply maz-rounded;
      }

      &.--rounded {
        @apply maz-rounded-full;
      }

      &.--xl {
        @apply maz-px-8 maz-text-xl;

        padding-top: 1.325rem;
        padding-bottom: 1.325rem;
      }

      &.--lg {
        @apply maz-px-6 maz-text-lg;

        padding-top: 1rem;
        padding-bottom: 1rem;
      }

      &.--md {
        @apply maz-px-4;

        padding-top: 0.88rem;
        padding-bottom: 0.88rem;
      }

      &.--sm {
        @apply maz-px-3 maz-text-sm;

        padding-top: 0.625rem;
        padding-bottom: 0.625rem;
      }

      &.--xs {
        @apply maz-px-2 maz-text-sm;

        padding-top: 0.3rem;
        padding-bottom: 0.3rem;
      }

      &.--mini {
        @apply maz-px-1 maz-text-xs;

        padding-top: 0.2rem;
        padding-bottom: 0.2rem;
      }

      transition:
        background 300ms ease-in-out 0ms,
        color 300ms ease-in-out 0ms;

      /* Not disabled */

      &:not(.--disabled) {
        &:hover,
        &:focus {
          @apply maz-bg-gray-200;
        }
      }

      /* Fab */

      &.--fab {
        @apply maz-flex maz-items-center
          maz-justify-center maz-rounded-full maz-px-0 maz-py-0;

        &:not(.--no-elevation) {
          @apply maz-elevation;
        }

        &.--xl {
          @apply maz-h-[4.125rem] maz-w-[4.125rem] maz-text-xl;
        }

        &.--lg {
          @apply maz-h-[3.375rem] maz-w-[3.375rem] maz-text-lg;
        }

        &.--md {
          @apply maz-h-12 maz-w-12;
        }

        &.--sm {
          @apply maz-h-[2.375rem] maz-w-[2.375rem] maz-text-sm;
        }

        &.--xs {
          @apply maz-h-[1.725rem] maz-w-[1.725rem] maz-text-sm;
        }

        &.--mini {
          @apply maz-h-[1.4rem] maz-w-[1.4rem] maz-text-xs;
        }
      }

      &.--block {
        @apply maz-w-full;
      }

      &.--primary {
        @apply maz-bg-primary maz-text-primary-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-primary-600;
        }
      }

      &.--secondary {
        @apply maz-bg-secondary maz-text-secondary-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-secondary-600;
        }
      }

      &.--info {
        @apply maz-bg-info maz-text-info-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-info-600;
        }
      }

      &.--success {
        @apply maz-bg-success maz-text-success-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-success-600;
        }
      }

      &.--warning {
        @apply maz-bg-warning maz-text-warning-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-warning-600;
        }
      }

      &.--danger {
        @apply maz-bg-danger maz-text-white;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-danger-600;
        }
      }

      &.--white {
        @apply maz-bg-white maz-text-white-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-gray-300;
        }
      }

      &.--black {
        @apply maz-bg-black maz-text-black-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-gray-800;
        }
      }

      &.--transparent {
        @apply maz-bg-transparent;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-[#000]/[0.08] dark:maz-bg-[#FFF]/[0.08];
        }
      }

      &.--theme {
        @apply maz-bg-theme maz-text-color;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-theme-hover;
        }
      }

      /* OUTLINE */

      &.--primary-outline {
        @apply maz-border-primary maz-bg-primary-alpha-05 maz-text-primary;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-primary maz-text-primary-contrast;
        }
      }

      &.--secondary-outline {
        @apply maz-border-secondary maz-bg-secondary-alpha-05 maz-text-secondary;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-secondary maz-text-white;
        }
      }

      &.--info-outline {
        @apply maz-border-info maz-bg-info-alpha-05 maz-text-info;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-info maz-text-white;
        }
      }

      &.--success-outline {
        @apply maz-border-success maz-bg-success-alpha-05 maz-text-success;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-success maz-text-white;
        }
      }

      &.--danger-outline {
        @apply maz-border-danger maz-bg-danger-alpha-05 maz-text-danger;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-danger maz-text-white;
        }
      }

      &.--warning-outline {
        @apply maz-border-warning maz-bg-warning-alpha-05 maz-text-warning;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-warning maz-text-white;
        }
      }

      &.--white-outline {
        @apply maz-border-white maz-text-white;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-white maz-text-white-contrast;
        }
      }

      &.--black-outline {
        @apply maz-border-black maz-text-black;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-black maz-text-black-contrast;
        }
      }

      &.--theme-outline {
        @apply maz-border-theme maz-text-theme;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-theme maz-text-color;
        }
      }

      /* PASTEL */

      &.--primary-pastel {
        @apply maz-bg-primary-50 maz-text-primary;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-primary maz-text-primary-contrast;
        }
      }

      &.--secondary-pastel {
        @apply maz-bg-secondary-50 maz-text-secondary;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-secondary maz-text-secondary-contrast;
        }
      }

      &.--info-pastel {
        @apply maz-bg-info-50 maz-text-info;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-info maz-text-info-contrast;
        }
      }

      &.--success-pastel {
        @apply maz-bg-success-50 maz-text-success;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-success maz-text-success-contrast;
        }
      }

      &.--danger-pastel {
        @apply maz-bg-danger-50 maz-text-danger;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-danger maz-text-danger-contrast;
        }
      }

      &.--warning-pastel {
        @apply maz-bg-warning-50 maz-text-warning;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-warning maz-text-warning-contrast;
        }
      }

      &.--white-pastel {
        @apply maz-bg-gray-100 maz-text-white;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-white maz-text-white-contrast;
        }
      }

      &.--black-pastel {
        @apply maz-bg-gray-200 maz-text-black;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-black maz-text-black-contrast;
        }
      }

      &.--theme-pastel {
        @apply maz-bg-gray-200 maz-text-black;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-black maz-text-black-contrast;
        }
      }

      /* DISABLED */

      &.--disabled {
        @apply maz-cursor-not-allowed maz-bg-color-lighter maz-text-gray-400;
      }

      &.--loading {
        @apply maz-cursor-wait;
      }

      &.--no-padding {
        @apply maz-p-0;
      }
    }
  }
</style>
