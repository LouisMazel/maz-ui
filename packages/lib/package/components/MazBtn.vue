<template>
  <Component
    v-bind="$attrs"
    :is="component"
    :disabled="isDisabled"
    class="m-btn"
    :class="[
      `--${size}`,
      btnColorClass,
      cursorClass,
      variantClass,
      {
        '--block': block,
        '--no-underline': noUnderline,
        '--no-leading': noLeading,
        '--fab': fab,
        '--loading': loading,
        '--disabled': isDisabled,
        '--icon': hasIcon,
        '--rounded': rounded,
      },
    ]"
    :type="btnType"
  >
    <div v-if="hasLeftIcon" class="m-btn__icon-left">
      <slot name="left-icon">
        <MazIcon v-if="leftIcon" :name="leftIcon" />
      </slot>
    </div>
    <span>
      <slot></slot>
    </span>
    <div v-if="hasRightIcon" class="m-btn__icon-right">
      <slot name="right-icon">
        <MazIcon v-if="rightIcon" :name="rightIcon" />
      </slot>
    </div>
    <div
      v-if="hasLoader"
      class="m-btn__loading-wrapper"
      :class="loaderBgColorClass"
    >
      <MazSpinner size="2em" :color="loaderColor" />
    </div>
  </Component>
</template>

<script lang="ts">
  export type { Color, Size } from './types'
</script>

<script lang="ts" setup>
  import { computed, PropType, useAttrs, useSlots } from 'vue'
  import MazSpinner from './MazSpinner.vue'
  import MazIcon from './MazIcon.vue'

  import { Color, Size } from './types'

  const { href, to } = useAttrs()
  const slots = useSlots()

  const props = defineProps({
    variant: {
      type: String,
      default: 'button',
      validator: (value: string) => {
        return ['button', 'link'].includes(value)
      },
    },
    size: {
      type: String as PropType<Size>,
      default: 'md',
      validator: (value: string) => {
        return ['mini', 'xs', 'sm', 'md', 'lg', 'xl'].includes(value)
      },
    },
    color: {
      type: String as PropType<Color>,
      default: 'primary',
      validator: (value: Color) => {
        return [
          'primary',
          'secondary',
          'info',
          'success',
          'warning',
          'danger',
          'white',
          'black',
          'transparent',
        ].includes(value)
      },
    },
    type: {
      type: String,
      default: 'button',
      validator: (value: string) => {
        return ['button', 'submit'].includes(value)
      },
    },
    rounded: { type: Boolean, default: false },
    outline: { type: Boolean, default: false },
    pastel: { type: Boolean, default: false },
    block: { type: Boolean, default: false },
    noUnderline: { type: Boolean, default: false },
    noLeading: { type: Boolean, default: false },
    loading: { type: Boolean, default: false },
    disabled: { type: Boolean, default: false },
    fab: { type: Boolean, default: false },
    leftIcon: { type: String, default: undefined },
    rightIcon: { type: String, default: undefined },
  })

  const component = computed(() => {
    if (href) return 'a'
    else if (to) return 'router-link'
    return 'button'
  })

  const btnColorClass = computed(() =>
    props.pastel
      ? `--${props.color}-pastel`
      : props.outline || props.loading
      ? `--${props.color}-outline`
      : `--${props.color}`,
  )
  const isDisabled = computed(
    () => (props.loading || props.disabled) && component.value === 'button',
  )
  const cursorClass = computed(() =>
    isDisabled.value ? '--cursor-default' : '--cursor-pointer',
  )
  const variantClass = computed(() => `--is-${props.variant}`)
  const loaderBgColorClass = computed(() => `--${props.color}`)
  const loaderColor = computed(() =>
    ['white'].includes(props.color) ? 'black' : 'white',
  )
  const hasLoader = computed(() => props.loading && props.variant === 'button')
  const hasLeftIcon = computed(() => !!slots['left-icon'] || props.leftIcon)
  const hasRightIcon = computed(() => !!slots['right-icon'] || props.rightIcon)
  const hasIcon = computed(() => hasLeftIcon.value || hasRightIcon.value)
  const btnType = computed(() =>
    component.value === 'button' ? props.type : undefined,
  )
</script>

<style lang="postcss" scoped>
  .m-btn {
    @apply maz-border maz-border-solid maz-border-transparent maz-text-center maz-text-base;

    & span {
      @apply maz-leading-none;
    }

    &__icon-left {
      @apply maz-mr-2 maz--ml-1 maz-leading-none;
    }

    &__icon-right {
      @apply maz-ml-2 maz--mr-1 maz-leading-none;
    }

    &.--cursor-pointer {
      @apply maz-cursor-pointer;
    }

    &.--cursor-default {
      @apply maz-cursor-default;
    }

    &.--is-link {
      @apply maz-outline-none maz-inline-flex maz-items-center maz-bg-transparent maz-text-normal-text maz-transition maz-duration-200 maz-ease-in-out;

      &:not(.--no-leading) {
        @apply maz-leading-9;
      }

      &:not(:disabled):hover,
      &:not(:disabled):focus {
        &:not(.--no-underline) {
          @apply maz-underline;
        }
      }

      &.--secondary {
        @apply maz-text-secondary;
      }

      &.--info {
        @apply maz-text-info;
      }

      &.--warning {
        @apply maz-text-warning-600;
      }

      &.--danger {
        @apply maz-text-danger-600;
      }

      &.--success {
        @apply maz-text-success-600;
      }

      &.--white {
        @apply maz-text-white;
      }

      &.--black {
        @apply maz-text-black;
      }
    }

    &.--is-button {
      @apply maz-relative maz-inline-flex maz-items-center maz-justify-center maz-overflow-hidden
      maz-rounded-lg maz-border-transparent maz-bg-transparent maz-font-base maz-font-medium maz-no-underline
      maz-transition maz-duration-300 maz-ease-in-out;

      &.--rounded {
        @apply maz-rounded-full;
      }

      &.--xl {
        @apply maz-px-8 maz-text-xl;

        padding-top: 1.325rem;
        padding-bottom: 1.325rem;
      }

      &.--lg {
        @apply maz-px-6 maz-text-lg;

        padding-top: 1rem;
        padding-bottom: 1rem;
      }

      &.--md {
        @apply maz-px-4 maz-text-base;

        padding-top: 0.88rem;
        padding-bottom: 0.88rem;
      }

      &.--sm {
        @apply maz-px-3 maz-text-sm;

        padding-top: 0.625rem;
        padding-bottom: 0.625rem;
      }

      &.--xs {
        @apply maz-px-2 maz-text-sm;

        padding-top: 0.3rem;
        padding-bottom: 0.3rem;
      }

      &.--mini {
        @apply maz-px-1 maz-text-xs;

        padding-top: 0.2rem;
        padding-bottom: 0.2rem;
      }

      &.--icon {
        @apply maz-py-2;
      }

      transition: background 300ms ease-in-out 0ms, color 300ms ease-in-out 0ms;

      /* Not disabled */

      &:not(.--disabled) {
        &:hover,
        &:focus {
          @apply maz-bg-gray-200;
        }
      }

      /* Fab */

      &.--fab {
        @apply maz-flex maz-h-12 maz-w-12 maz-items-center maz-justify-center maz-rounded-full
        maz-px-0 maz-py-0 maz-drop-shadow-md;
      }

      &.--block {
        @apply maz-w-full;
      }

      &.--primary {
        @apply maz-bg-primary maz-text-primary-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-primary-600;
        }
      }

      &.--secondary {
        @apply maz-bg-secondary maz-text-secondary-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-secondary-600;
        }
      }

      &.--info {
        @apply maz-bg-info maz-text-info-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-info-600;
        }
      }

      &.--success {
        @apply maz-bg-success maz-text-success-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-success-600;
        }
      }

      &.--warning {
        @apply maz-bg-warning maz-text-warning-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-warning-600;
        }
      }

      &.--danger {
        @apply maz-bg-danger maz-text-white;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-danger-600;
        }
      }

      &.--white {
        @apply maz-bg-white maz-text-white-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-gray-300;
        }
      }

      &.--black {
        @apply maz-bg-black maz-text-black-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-gray-800;
        }
      }

      &.--transparent {
        @apply maz-bg-transparent maz-text-normal-text;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-color-light;
        }
      }

      &.--primary-outline {
        @apply maz-border-primary maz-text-primary;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-primary maz-text-primary-contrast;
        }
      }

      &.--secondary-outline {
        @apply maz-border-secondary maz-text-secondary;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-secondary maz-text-white;
        }
      }

      &.--info-outline {
        @apply maz-border-info maz-text-info;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-info maz-text-white;
        }
      }

      &.--success-outline {
        @apply maz-border-success maz-text-success;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-success maz-text-white;
        }
      }

      &.--danger-outline {
        @apply maz-border-danger maz-text-danger;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-danger maz-text-white;
        }
      }

      &.--warning-outline {
        @apply maz-border-warning maz-text-warning;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-warning maz-text-white;
        }
      }

      &.--white-outline {
        @apply maz-border-white maz-text-white;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-white maz-text-white;
        }
      }

      &.--black-outline {
        @apply maz-border-black maz-text-black;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-black maz-text-black;
        }
      }

      &.--primary-pastel {
        @apply maz-bg-primary-50 maz-text-primary;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-primary maz-text-primary-contrast;
        }
      }

      &.--secondary-pastel {
        @apply maz-bg-secondary-50 maz-text-secondary;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-secondary maz-text-secondary-contrast;
        }
      }

      &.--info-pastel {
        @apply maz-bg-info-50 maz-text-info;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-info maz-text-info-contrast;
        }
      }

      &.--success-pastel {
        @apply maz-bg-success-50 maz-text-success;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-success maz-text-success-contrast;
        }
      }

      &.--danger-pastel {
        @apply maz-bg-danger-50 maz-text-danger;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-danger maz-text-danger-contrast;
        }
      }

      &.--warning-pastel {
        @apply maz-bg-warning-50 maz-text-warning;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-warning maz-text-warning-contrast;
        }
      }

      &.--white-pastel {
        @apply maz-bg-gray-100 maz-text-white;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-white maz-text-white-contrast;
        }
      }

      &.--black-pastel {
        @apply maz-bg-gray-200 maz-text-black;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-black maz-text-black-contrast;
        }
      }

      /* Disabled */
      &.--disabled {
        @apply maz-cursor-not-allowed maz-bg-color-light maz-text-gray-400;
      }

      /* Loader */
      .m-btn__loading-wrapper {
        @apply maz-absolute maz-inset-0 maz-flex maz-items-center maz-justify-center;

        &.--primary {
          @apply maz-bg-primary;
        }

        &.--secondary {
          @apply maz-bg-secondary;
        }

        &.--info {
          @apply maz-bg-info;
        }

        &.--warning {
          @apply maz-bg-warning;
        }

        &.--success {
          @apply maz-bg-success;
        }

        &.--danger {
          @apply maz-bg-danger;
        }

        &.--white {
          @apply maz-bg-white;
        }

        &.--black {
          @apply maz-bg-black;
        }
      }
    }
  }
</style>
