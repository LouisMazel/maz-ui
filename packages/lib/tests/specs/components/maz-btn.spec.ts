import MazBtn from '@package/components/MazBtn.vue'
import { shallowMount } from '@vue/test-utils'

describe('MazBtn', () => {
  test('Should match wtesth the snapshot', () => {
    const wrapper = shallowMount(MazBtn)
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('renders the component elements and classes correctly', () => {
    const wrapper = shallowMount(MazBtn)

    expect(wrapper.find('button').exists()).toBe(true)
    expect(wrapper.find('.m-btn__icon-left').exists()).toBe(false)

    expect(wrapper.find('span').exists()).toBe(true)
    expect(wrapper.find('.m-btn__icon-right').exists()).toBe(false)
    expect(wrapper.find('.m-btn__loading-wrapper').exists()).toBe(false)

    expect(wrapper.find('.m-btn').exists()).toBe(true)
  })

  test('renders the button content correctly using slots', () => {
    const wrapper = shallowMount(MazBtn, {
      slots: {
        default: 'Button Text',
        'left-icon': '<div class="left-icon">Left Icon</div>',
        'right-icon': '<div class="right-icon">Right Icon</div>',
      },
    })

    expect(wrapper.text()).toContain('Button Text')
    expect(wrapper.find('.m-btn__icon-left').exists()).toBe(true)
    expect(wrapper.find('.m-btn__icon-right').exists()).toBe(true)
  })

  test('shows the loading spinner when the "loading" prop is set to "true"', () => {
    const wrapper = shallowMount(MazBtn, {
      propsData: {
        loading: true,
      },
    })

    // Vérifie que le wrapper de chargement est affiché
    expect(wrapper.find('.m-btn__loading-wrapper').exists()).toBe(true)
  })
})
