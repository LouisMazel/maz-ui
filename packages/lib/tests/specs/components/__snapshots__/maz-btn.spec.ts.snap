// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MazBtn > Should match with the snapshot 1`] = `
"<button data-v-fa7305fb=\\"\\" class=\\"m-btn --md --primary --cursor-pointer --is-button\\" type=\\"button\\">
  <!--v-if--><span data-v-fa7305fb=\\"\\" class=\\"maz-flex maz-flex-center\\"></span>
  <!--v-if-->
  <!--v-if-->
</button>"
`;
